command.go

```diff
@@ -491,6 +468,30 @@ Loop:
 	return commands
 }
 
+func hasDefaultValue(name string, fs *flag.FlagSet) bool {
+	flag := fs.Lookup(name)
+	if flag == nil {
+		return false
+	}
+	return flag.NoOptDefVal != ""
+}
+
+func isShortFlagWithNoDefaultValue(arg string, flags *flag.FlagSet) bool {
```

Comment:
I think this func `isShortFlagWithNoDefaultValue` is confusing.

*partly copied from pr 615* 
``` go
func hasDefaultShort_s(arg string, c *Command) bool { // should probably return (bool, error)
	if strings.HasPrefix(arg, "-") {
		if !strings.Contains(arg, "=") {
                       // needs to change to a range to deal with -abcd goldfish
			if len(arg) == 2 {
				flag := c.Flags().ShorthandLookup(arg[1:])
				if flag == nil {
					return false
				}
				//shortHasNoOptDefVal
				return flag.NoOptDefVal != ""
			}
		}
  	}
	return false
 }
```
It's also a good place to solve the `-abcd goldfish` issue.

I think some of the issues and confusion are created as this code can't be sure if a flag exists and have to choose a default. Is it possible to

1. change these helper funcs to return `(bool, error)`, or
1. check the flags before hand (might be hard as this code is figure out which flags need checking)

A lot of the `if flag == nil {` and `if len(name) == 0 {` code could be removed making this code clearer.

Reply:
> I think this func isShortFlagWithNoDefaultValue is confusing.

Sorry, I don't understand why you feel it is confusing. Does the name not reflect the behaviour of the function?

> change these helper funcs to return (bool, error)

What would be the error case?

> check the flags before hand

This is the code which checks flag, I don't understand what we could do be fore this.

> A lot of the if flag == nil { and if len(name) == 0 { code could be removed making this code clearer.

`if len(name) == 0` can be removed I think, ys. `flag == nil` is still important. It's what tells us if this is a flag or an arg.

Reply:
If you would rather I omit the refactoring, and only add the test case for now, I'm fine with that.